File: src/core/rendering/conversion/ninja_animation.ts
import {
    AnimationClip,
    Euler,
    InterpolateLinear,
    InterpolateSmooth,
    KeyframeTrack,
    Quaternion,
    QuaternionKeyframeTrack,
    VectorKeyframeTrack,
} from "three";
import { NjModel, NjObject } from "../../data_formats/parsing/ninja";
import {
    NjInterpolation,
    NjKeyframeTrackType,
    NjMotion,
} from "../../data_formats/parsing/ninja/motion";

export const PSO_FRAME_RATE = 30;

export function create_animation_clip(
    nj_object: NjObject<NjModel>,
    nj_motion: NjMotion,
): AnimationClip {
    const interpolation =
        nj_motion.interpolation === NjInterpolation.Spline ? InterpolateSmooth : InterpolateLinear;

    const tracks: KeyframeTrack[] = [];

    nj_motion.motion_data.forEach((motion_data, bone_id) => {
        const bone = nj_object.get_bone(bone_id);
        if (!bone) return;

        motion_data.tracks.forEach(({ type, keyframes }) => {
            const times: number[] = [];
            const values: number[] = [];

            for (const keyframe of keyframes) {
                times.push(keyframe.frame / PSO_FRAME_RATE);

                if (type === NjKeyframeTrackType.Rotation) {
                    const order = bone.evaluation_flags.zxy_rotation_order ? "ZXY" : "ZYX";
                    const quat = new Quaternion().setFromEuler(
                        new Euler(keyframe.value.x, keyframe.value.y, keyframe.value.z, order),
                    );

                    values.push(quat.x, quat.y, quat.z, quat.w);
                } else {
                    values.push(keyframe.value.x, keyframe.value.y, keyframe.value.z);
                }
            }

            if (type === NjKeyframeTrackType.Rotation) {
                tracks.push(
                    new QuaternionKeyframeTrack(
                        `.bones[${bone_id}].quaternion`,
                        times,
                        values,
                        interpolation,
                    ),
                );
            } else {
                const name =
                    type === NjKeyframeTrackType.Position
                        ? `.bones[${bone_id}].position`
                        : `.bones[${bone_id}].scale`;

                tracks.push(new VectorKeyframeTrack(name, times, values, interpolation));
            }
        });
    });

    return new AnimationClip(
        "Animation",
        (nj_motion.frame_count - 1) / PSO_FRAME_RATE,
        tracks,
    ).optimize();
}

File: web/src/jsMain/kotlin/world/phantasmal/web/core/rendering/conversion/NinjaAnimation.kt
package world.phantasmal.web.core.rendering.conversion

import world.phantasmal.core.asArray
import world.phantasmal.core.jsArrayOf
import world.phantasmal.psolib.fileFormats.ninja.NinjaObject
import world.phantasmal.psolib.fileFormats.ninja.NjInterpolation
import world.phantasmal.psolib.fileFormats.ninja.NjKeyframeTrack
import world.phantasmal.psolib.fileFormats.ninja.NjMotion
import world.phantasmal.web.externals.three.*

const val PSO_FRAME_RATE: Int = 30
const val PSO_FRAME_RATE_DOUBLE: Double = PSO_FRAME_RATE.toDouble()

fun createAnimationClip(njObject: NinjaObject<*, *>, njMotion: NjMotion): AnimationClip {
    val interpolation =
        if (njMotion.interpolation == NjInterpolation.Spline) InterpolateSmooth
        else InterpolateLinear

    val tracks = jsArrayOf<KeyframeTrack>()

    for ((boneIndex, motionData) in njMotion.motionData.withIndex()) {
        val bone = njObject.getBone(boneIndex) ?: continue

        for (track in motionData.tracks) {
            if (track.keyframes.isEmpty()) continue

            val baseName = ".bones[$boneIndex]"
            val times = jsArrayOf<Double>()

            for (keyframe in track.keyframes) {
                times.push(keyframe.frame / PSO_FRAME_RATE_DOUBLE)
            }

            val values = jsArrayOf<Double>()

            tracks.push(
                when (track) {
                    is NjKeyframeTrack.Position -> {
                        for (keyframe in track.keyframes) {
                            values.push(
                                keyframe.value.x.toDouble(),
                                keyframe.value.y.toDouble(),
                                keyframe.value.z.toDouble(),
                            )
                        }

                        VectorKeyframeTrack(
                            "$baseName.position",
                            times.asArray(),
                            values.asArray(),
                            interpolation,
                        )
                    }

                    is NjKeyframeTrack.EulerAngles -> {
                        for (keyframe in track.keyframes) {
                            val quat = Quaternion().setFromEuler(
                                Euler(
                                    keyframe.value.x.toDouble(),
                                    keyframe.value.y.toDouble(),
                                    keyframe.value.z.toDouble(),
                                    if (bone.evaluationFlags.zxyRotationOrder) "ZXY" else "ZYX"
                                ),
                            )

                            values.push(quat.x, quat.y, quat.z, quat.w)
                        }

                        QuaternionKeyframeTrack(
                            "$baseName.quaternion",
                            times.asArray(),
                            values.asArray(),
                            interpolation,
                        )
                    }

                    is NjKeyframeTrack.Scale -> {
                        for (keyframe in track.keyframes) {
                            values.push(
                                keyframe.value.x.toDouble(),
                                keyframe.value.y.toDouble(),
                                keyframe.value.z.toDouble(),
                            )
                        }

                        VectorKeyframeTrack(
                            "$baseName.scale",
                            times.asArray(),
                            values.asArray(),
                            interpolation,
                        )
                    }

                    is NjKeyframeTrack.Quaternion -> {
                        for (keyframe in track.keyframes) {
                            values.push(
                                keyframe.imaginary.x.toDouble(),
                                keyframe.imaginary.y.toDouble(),
                                keyframe.imaginary.z.toDouble(),
                                keyframe.real.toDouble(),
                            )
                        }

                        QuaternionKeyframeTrack(
                            "$baseName.quaternion",
                            times.asArray(),
                            values.asArray(),
                            interpolation,
                        )
                    }
                }
            )
        }
    }

    return AnimationClip(
        "Animation",
        (njMotion.frameCount - 1) / PSO_FRAME_RATE_DOUBLE,
        tracks.asArray(),
    ).optimize()
}
